Upgrade Resolver for gears, analyses and lookup

ehlertjd opened the pull request:
Resolver now uses ContainerStorage for projections and tree knowledge. Resolver also uses DB filtering to find the next node, rather than retrieving all children then filtering. 

I've added indexes for [`parent`, `label`] to projects, sessions and acquisitions. This caused intermittent breakages in a few integration tests that were relying on DB returning records in insert-order - I believe that these are all fixed. Should I include a migration closure to drop the obsoleted indexes?

This implements virtual nodes for `gears`, `analyses`, and `files`. By default we return all children of a node, but you can select just the files or analyses by adding `files` or `analyses` to the path. **e.g.** resolving `scitran/Neuroscience/session-01` will return all acquisitions, analyses and files as children, but resolving `scitran/Neuroscience/session-01/files` will return just the list of files. This resolves the filename ambiguity that is introduced in #1089. 

Gears can be retrieved by using a path of `/gears/gear-name` or `/gears/<id:gear-id>`.

In addition, this adds a new `/api/lookup` endpoint that takes the same input as `/api/resolve`, but redirects to the appropriate GET handler for the resolved node. For example, performing a lookup of `scitran/Neuroscience/session-01` will redirect (transparently on the server side) to the GET handler for `/api/sessions/<session-01.id>`. The one exception is if you perform a lookup on a file node, it will simply return the file node (without an info object). 

Finally, in order to support the resolver/lookup in SDK2, polymorphism was added to `resolver.json` and a node-type defined for each type. This means that we need to set a `container_type` attribute in any response returned from `/api/resolve` or `/api/lookup`. In the case of lookup, this is done by setting a request environment variable `fw_container_type` and invoking the new utility function: `util.add_container_type`.

### Breaking Changes

* You can no longer resolve files by naming them directly. For example: `/scitran/Neuroscience/ses-01/acq/scan.nii.gz` **MUST** become `/scitran/Neuroscience/ses-01/acq/files/scan.nii.gz`.
* Renamed `node_type` to `container_type` to be consistent with other places where we identify the container type (e.g. search)

### Review Checklist

- Tests were added to cover all code changes
- Documentation was added / updated
- Code and tests follow standards in CONTRIBUTING.md

ehlertjd committed:
Added step to simplify swagger for code generation. Fixed a few missing types in schemas, added schemas for login/logout responses

ehlertjd committed:
Added additional model simplifications

ehlertjd committed:
Alias primitive types in definitions

ehlertjd committed:
Allow template parameters to be optional

Parameters can now be optional or required, which allows for more advanced template usage.
Updated existing templates to set the required flag.
Added documentation for root level analyses apis via existing templates.

ehlertjd committed:
Added support for conversion from patternProperties to additionalProperties in schema transpiler

ehlertjd committed:
Replace pure references in definitions with aliases

ryansanford committed:
Initial addition of reaper upload API doc

davidfarkas93 committed:
Add jobs/logs, jobs/prepare-complete, jobs/accept-failed-output endpoints doc

ambrussimon committed:
Document site rules

ehlertjd committed:
Checkpoint commit of modifications for codegen.

ehlertjd committed:
Add missing DELETE file

ehlertjd committed:
Reorder tags

ehlertjd committed:
Add schemas for gear APIs.

ehlertjd committed:
Add schema and endpoint for jobs

Added input schema for adding a job, and the endpoint for GET /job/{JobId}/logs.

ehlertjd committed:
Added x-sdk-include-empty for Gear model

ehlertjd committed:
Rename file to file-entry (python codegen had a problem with that name)

ehlertjd committed:
API Doc changes for codegen

ehlertjd committed:
Added date-time format to created/modified timestamps

ehlertjd committed:
Add docs for modify info

ehlertjd committed:
Set default values for info_exists.

ehlertjd committed:
Add documentation for PUT on file endpoints

ehlertjd committed:
Fix failing unit tests due to schema changes

ehlertjd committed:
Add missing parameter to collection/acquisitions

Added "session" query parameter that limits the returned acquisitions.

ehlertjd committed:
Add missing gear invocation endpoint

This endpoint returns the JSON schema for gear configuration.

ehlertjd committed:
Rename update_job endpoint

ehlertjd committed:
Add get_analyses endpoint documentation

This documents the endpoints for /<container>/<cid>/analyses as well as
/<container>/<cid>/<subcontainer>/analyses.

ehlertjd committed:
Fix failing tests for analyses

ehlertjd committed:
Update file to file-entry in analysis list

ehlertjd committed:
Update schemas for analyses in SDK

This update forces us to use the inflated job for single analysis end points.

ehlertjd committed:
Ignore '+' and '-' properties for download filters

ehlertjd committed:
Add schemas for batch job scheduling

Also refactored a few schemas to better match the SDK.

ehlertjd committed:
Add schema for adding job logs

ehlertjd committed:
Add API documentation for search endpoint

ehlertjd committed:
Add default values for search parameters

ehlertjd committed:
Cleanup packfile endpoints

Removed packfile endpoints for sessions, collections, and acquisitions.
Added missing parameters to packfile and packfile-end endpoints.

ehlertjd committed:
Update packfile output definition

ehlertjd committed:
Add model merge for swagger codegen

This commit will allow merging input and output models based on
the `x-sdk-model` property in the JSON definitions.

ehlertjd committed:
Add documentation for resolver endpoint

Closes #1088

ehlertjd committed:
Add additional handling for polymorphic models

ehlertjd committed:
Generalize resolver approach

This commit updates resolver to make use of mongo indexes and retrieve
fewer nodes from the database. Also allows resolving just a single
container id without retrieving children.

ehlertjd committed:
Add lookup endpoint that routes to GET for path

In order to support polymorphism in clients, this commit also adds
a new optional request environment variable `fw_node_type` which,
when set should be included as `node_type` in the result.

ehlertjd committed:
Move list_projection into ContainerStorage

ehlertjd committed:
Use ContainerStorage for resolver

ehlertjd committed:
Add swagger documentation for lookup endpoint

ehlertjd committed:
Fix test failing due to indeterminate ordering

ehlertjd committed:
Refactor resolver to support virtual nodes

ehlertjd committed:
Add gears to resolver

This does not yet support gear versioning.

ehlertjd committed:
Add optional query to get_analyses

ehlertjd committed:
Add analyses to resolver

ehlertjd committed:
Add resolver definitions for gears and analyses

ehlertjd committed:
Improve resolver test coverage

ehlertjd committed:
Rename resolver's node_type to container_type

kofalt commented on api/resolver.py:
```diff
@@ -1,177 +1,361 @@
+
+        # Check for analyses
+        if path_peek(path_in) == 'analyses':
+            if self.analyses:
```
I think these two conditionals should be combined and the `raise` removed. Right now, there's nothing stopping someone calling a container `analyses` (ref https://github.com/scitran/core/pull/1089#issuecomment-370565314) and I would expect a not-found error to occur if nothing is found.

kofalt commented on api/resolver.py:
```diff
@@ -180,47 +364,34 @@ class Resolver(object):
-    @staticmethod
-    def resolve(path):
-
+    def resolve(self, path):
```
This function is pretty slick relative to the previous one. I'm a little iffy on the callees modifying the passed variables, but I can see the advantage.

kofalt commented on api/resolver.py:
```diff
@@ -180,47 +364,34 @@ class Resolver(object):
+        return {
+            'path': resolved_path,
+            'children': resolved_children
+        }
```
Overall, this file roughly doubled in length. It definitely gained some features and documentation, but I think lost a lot of readability along the way. Is there anything we can do to improve that?

ehlertjd replied:
I think your review had a lot of good suggestions to help with that, but I'll keep an eye out for other improvements that can be made.

ehlertjd commented on api/dao/containerstorage.py:
```diff
@@ -100,7 +100,12 @@ def recalc_sessions_compliance(self, project_id=None):
+            list_projection={'info': 0, 'analyses': 0, 'subject.firstname': 0,
+                'subject.lastname': 0, 'subject.sex': 0, 'subject.age': 0,
+                'subject.race': 0, 'subject.ethnicity': 0, 'subject.info': 0,
+                'files.info': 0, 'tags': 0})
```
Alternatively, subclasses could just override the `get_list_projection` method? I agree that the above is awkward.

ehlertjd commented on api/handlers/grouphandler.py:
```diff
@@ -22,6 +22,7 @@ def get(self, _id):
             self._filter_permissions([result], self.uid)
         if self.is_true('join_avatars'):
             ContainerHandler.join_user_info([result])
+        util.add_container_type(self.request, result)
```
It will only show up if you set the `fw_container_type` key in the environment, so currently it's only for forwarded requests. I wouldn't have any complaints about that key showing up everywhere, so I'm open to that conversation; if that were the case I'd advocate moving the logic into `ContainerStorage`.

ehlertjd commented on api/resolver.py:
```diff
@@ -1,177 +1,361 @@
+
+The goal of the resolver is to provide a virtual graph that can be navigated using
+path notation. Below is how the graph will ultimately be represented. Currently
+subjects are not formalized and are excluded from the implementation.
```
My understanding is that @nagem is planning on formalizing subjects in the very near future. If that's not the case (and maybe even if it is) I think that it's a great time to add subjects. If we add them before formalizing them, I think we'd need (or want to) duplicate the merge logic for subjects.

ehlertjd commented on api/resolver.py:
```diff
@@ -1,177 +1,361 @@
+    if not path_in:
+        return False, None
+
+    value = path_in.popleft()
```
Nope, that's just taking the next value off the deque before parsing it and returning it. The input requirements haven't changed.

ehlertjd commented on api/handlers/resolvehandler.py:
```diff
@@ -1,28 +1,81 @@
+        # The client will depend on container_type being set so that it can map to the correct type
+        destination_environ['REQUEST_METHOD'] = 'GET'
+        destination_environ['fw_container_type'] = dest['container_type']
+        destination_request = Request(destination_environ)
```
I agree - my intent was to move this into a utility function once we're able to merge both PRs

ehlertjd commented on api/resolver.py:
```diff
@@ -1,177 +1,361 @@
+        
+        child = results[0]
+
+        self.storage.filter_deleted_files(child)
```
I'm not aware of any way to use a projection to filter elements from an array - I think we would have to use the aggregation framework.

ehlertjd commented on api/resolver.py:
```diff
@@ -1,177 +1,361 @@
+        query = {}
+        if criterion:
+            if use_id or self.use_id:
+                if self.storage.use_object_id:
```
`use_id` is set if the user specifies `<id:some_id>` as the criterion, `self.use_id` is for the group container case where we always search on id. Not sure I get the second half of your question - the next if block is to determine the format of id field. Arguably, that could be another method on ContainerStorage `mongo_id_from_string` or something like that.

ehlertjd commented on api/resolver.py:
```diff
@@ -1,177 +1,361 @@
+        use_id, criterion = parse_criterion(path_in)
+        parent = get_parent(path_out)
+        # Peek to see if we need files for the next path element
+        fetch_files = (path_peek(path_in) in ['files', None])
```
I think this goes away once we formalize files as their own collection. As long as files exist on containers I don't think it's inappropriate for container-centric code to deal with that fact. In this particular case the goal is to try to keep what we're retrieving from the database to what we actually need.

ehlertjd commented on api/dao/containerstorage.py:
```diff
@@ -291,24 +297,28 @@ def get_all_for_targets(self, target_type, target_ids, user=None, projection=Non
-        analyses = self.get_all_el({'parent.type': parent_type, 'parent.id': parent_id}, None, None)
+    def get_analyses(self, query, parent_type, parent_id, inflate_job_info=False, projection=None, **kwargs):
+        if query is None:
+            query = {}
```
`ContainerStorage.get_all_ell` allows None for query, so I was following that model.

ehlertjd committed:
Resolve review comments

Overall, increase documentation and attempt to reduce complexity in
resolver.py.
Also changed the paradigm for ContainerStorage list projections to use
function overrides.

ehlertjd commented:
@kofalt I believe I've addressed most of your comments. resolver.py did get a little simpler:

## Before
```
github.com/AlDanial/cloc v 1.76  T=1.03 s (1.0 files/s, 387.0 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           1             86             85            226
-------------------------------------------------------------------------------
```

## After
```
github.com/AlDanial/cloc v 1.76  T=1.05 s (1.0 files/s, 373.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           1             90            104            198
-------------------------------------------------------------------------------
```
